### ISSUE TITLE
Authentication in Black-Box Testing

### ISSUE THREAD

**Issue by qixin5** [EMResearch/EvoMaster#267]
Hi EvoMaster authors,

"Authentication: we support auth based on authentication headers and cookies. However, at the moment we do not support OAuth."

I wonder does EvoMaster currently support any personal-access-token-based authentication? Thanks.

**Comment by arcuri82** [comment]
hi,
 what you mean with `personal-access-token-based authentication`, exactly? can u give us some examples? (if it is a token that can be pre-generated, and passed in a HTTP header, then it should be doable)

**Comment by qixin5** [comment]
Hi @arcuri82,

Yes, it is a token pre-generated. In curl command, I use `curl --header "PRIVATE-TOKEN: MY-TOKEN-HERE" ...`.

**Comment by arcuri82** [comment]
Hi,
 in that case, it should be doable, eg, by using:

```
 AuthenticationDto dto = new AuthenticationDto("some id of your choice");
 dto.headers.add(new HeaderDto("PRIVATE-TOKEN",  "MY-TOKEN-HERE"));
```

used as returned object in `getInfoForAuthentication()`

(But we need to provide better documentation to do it...)

**Comment by qixin5** [comment]
HI @arcuri82,

Is this for white-box testing? Should I create a controller class that extends SutController and implements `getInfoForAuthentication()`?

How about black-box testing?

Thanks.

**Comment by arcuri82** [comment]
Hi @qixin5 ,

hmmm, very good point... I was assuming WBT. Not BBT. We do not support authentication in BBT :(  I should make it clear in the documentation.

WBT is much better than BBT. So, you should do WBT if you can. Or are you testing a remote service you do not have direct access to?

**Comment by qixin5** [comment]
Hi @arcuri82,

I'm trying to test a locally hosted service with EvoMaster. The service however is not written in Java, so I have to use black-box testing ... But thanks.

**Comment by arcuri82** [comment]
@qixin5 in which programming language is it written? is it open-source by any chance?

**Comment by qixin5** [comment]
@arcuri82, I'm looking at GitLab (ee version). It's in Ruby on Rails. I'm using their docker container (https://hub.docker.com/r/gitlab/gitlab-ee/), which contains source code.

**Comment by arcuri82** [comment]
@qixin5 we are in the process to support JavaScript and .Net, besides the JVM. But there is no plan to support Ruby.

Regarding auth in black-box testing, that something we will support, but not any time soon

**Comment by qixin5** [comment]
@arcuri82, thanks for letting me know.

**Comment by jonasbg** [comment]
I think BBT should accept custom http-headers as argument for instance like this:
```
java -jar evomaster.jar --blackBox true --bbSwaggerUrl https://myapi.example.com/swagger/v0.5/swagger.json  --outputFormat CSHARP_XUNIT --maxTime 10m --header "Authorization: Bearer access_token" --header "Accept: application/json"
```

**Comment by arcuri82** [comment]
header arguments `header0`, `header1` and `header2` have now been added to the master branch, which should solve this problem.
they will be available in the next release (scheduled hopefully within 1-2 weeks)

**Comment by arcuri82** [comment]
this should be now fixed in the new `1.3.0` release
